mantidproject/mantid
# Fix Variable Q Binning from SANS User File

**Description of work.**
SANS User files contain a line describing Q binning, which can be of the form
L/Q q1,dq1, q2, dq2, ...., qn /LOG, where qi is the i^th bin boundary, dqi is the i^th step size, and /LOG is an optional tag to set the step sizes to logarithmic.
This work correctly parses the lines in the user file.

Additionally, this PR rearranges how errors are handled while loading a user file to make it easier to see which line is causing which issue, and stops catching a general `except Exception` without a stack trace (so we can now locate the causes of new bugs more easily)

**Report to:** Robert Dalgliesh ISIS

**To test:**
1. Interfaces -> SANS -> SANS v2
2. Load the attached user file
[failed_user_file.txt](https://github.com/mantidproject/mantid/files/2699448/failed_user_file.txt)
3. -> Settings (on the side) -> Q, Wavelength, Detector Limits (on the top) 
4. In the Q limits box, the Q1D rebin string should be `0.004,-0.08,0.3,-0.06,0.6,-0.08,1.0` and the step type set to `Variable`
5. In the user file, edit the line `L/Q 0.004,0.08, 0.3, 0.06, 0.6, 0.08,1.0 /LOG` to `L/Q 0.004,0.08, 0.3, 0.06, 0.6, 0.08,1.0 /LIN`
6. Repeat steps 2-4, all should be the same except the rebin string should contain no negative signs
7. Try changing the line to separate the numbers with a variety of commas and spaces. Repeat steps 2-4

Fixes #24379 
Fixes #24378 
Fixes #24391 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
